### ISSUE TITLE
 Lacking (?) of Spring Cloud Stream - Error Handling documentation

### ISSUE THREAD

**Issue by stavalfi** [spring-cloud/spring-cloud-stream#1333]
Related to https://github.com/spring-cloud/spring-cloud-stream/issues/849

First, I explicitly want to note that I consider my self as a beginner with Spring framework. 

* What I'm looking for:

	(1)  What are you trying to solve?
	> Spring Cloud Stream supports publishing error messages received by the Spring Integration global error channel. Error messages sent to the errorChannel can be published to a specific destination at the broker by configuring a binding for the outbound target named error.
	> 
	- Is it a failure which related to my logic or spring's logic? 
	- Will I be able to send my own errors?
	- What is the difference between an `error channel` and `channel destination`?
	- When I'm using reactive API - will I still need this solution? 
	
	(2) List of all classes, annotations, and properties which related to Spring Cloud Stream - Error Handling.
    
    - spring.cloud.stream.bindings.error.destination
    - ErrorMessage
    - MessagingException
    - MessagePublishingErrorHandler
    
    Is that all?
    
	(3) Explicitly write the relationship between every 2 classes/annotations, if exist, and explicitly write why there is a connection between them.
    
	(4)  An example for each class, annotation, and property. Also, for each property, explicitly write what can be/must be specified inside the consumer and the producer.

**Comment by olegz** [comment]
@stavalfi Documentation has been updated. In fact there is a whole new Error Handling section. Snapshots will be updated shortly but if you want to build from the master
```
$> mvnw clean install  -Pfull -DskipTests
$> open spring-cloud-stream-core-docs/target/contents/reference/htmlsingle/index.html
```
There may still be a few things missing, so let us know

**Comment by olegz** [comment]
@stavalfi here, the snapshots are updated. https://docs.spring.io/spring-cloud-stream/docs/Elmhurst.BUILD-SNAPSHOT/reference/htmlsingle/#spring-cloud-stream-overview-error-handling
Let us know if something is missing, unlcear

**Comment by stavalfi** [comment]
@olegz it looks great!

It will take me a couple of days to go over it. Meanwhile, I found a really minor problem - missing `}` in the example here: https://docs.spring.io/spring-cloud-stream/docs/Elmhurst.BUILD-SNAPSHOT/reference/htmlsingle/#_application_error_handling

```java
@StreamListener(Sink.INPUT) // destination name 'input.myGroup'
public void handle(Person value)
	throw new RuntimeException("BOOM!");
}
```

Secondly, (Maybe it is related to StackOverflow) you specified that 

> For each input binding, Spring Cloud Stream creates a dedicated error channel with the following semantics `<channel-name>.<group-name>.errors`.

I copy & paste the code sample:

```java
@EnableBinding(Sink.class)
@SpringBootApplication
public class SpringCloudStreamKinesisConsumerApplication {

	public static void main(String[] args) {
		SpringApplication.run(SpringCloudStreamKinesisConsumerApplication.class, args);
	}

	@StreamListener(Sink.INPUT) // destination name 'input.myGroup'
	public void handle(String value) {
		throw new RuntimeException("BOOM!");
	}

	@StreamListener(Processor.INPUT + ".myGroup.errors") //channel name 'input.myGroup.errors'
	public void error(Message<?> message) {
		System.out.println("Handling ERROR: " + message);
	}
}
```
```java
spring.cloud.stream.bindings.input.destination=stava2
spring.cloud.stream.bindings.input.group=myGroup
spring.cloud.stream.bindings.input.content-type=application/json
```

But I get the following runtime error:

```java
***************************
APPLICATION FAILED TO START
***************************

Description:

A component required a bean named 'input.myGroup.errors' that could not be found.


Action:

Consider defining a bean named 'input.myGroup.errors' in your configuration.
```

I don't think I'm understanding the provided `action`.

**Comment by olegz** [comment]
@stavalfi Note on he document where it explains the meaning of the word `input` in the `spring.cloud.stream.bindings.input.group=myGroup`. In you case you are binding to destination `stava2`, so you shoudl have an error channel `stava2.myGroup.errors`.
Let me know

**Comment by stavalfi** [comment]
@olegz I'm confused :) Can you please fix the code sample & application.properties?

**Comment by olegz** [comment]
@stavalfi Thanks for your persistence, there is indeed an error in the documentation. so I'll fix it.
But here is the short version:

`spring.cloud.stream.bindings.input.destination=stava2` - states _bind the internal binding named `input` (such as `Processor.INPUT` for example) to the broker destination named `stava2` + `<groupName>`_. 

As you can see, the `destination` is not enough. It also needs a group name. By default the `<groupName>` will be anonymous value, resulting in the final destination (queue) named something like `stava2.anonymous.2K37rb06Q6m2r51-SPIDDQ`. However you can also instruct the system to use specific name. To do that you provide `spring.cloud.stream.bindings.input.group=myGroup`. This will result in the queue named `stava2.myGroup`. NOTE, it will either create such destination or, if it already exist, will simply bind to it.

Each binding will also have a dedicated error channel named `<destinationName>.<groupName>.errors` which in your case means `stava2.myGroup.errors`. The doc currently states `input.myGroup.errors` which is indeed wrong.

Hope this clarifies. 
Will update the docs.

**Comment by stavalfi** [comment]
I tried that but still getting the same error.

```java
@EnableBinding(Sink.class)
@SpringBootApplication
public class SpringCloudStreamKinesisConsumerApplication {

	public static void main(String[] args) {
		SpringApplication.run(SpringCloudStreamKinesisConsumerApplication.class, args);
	}

	@StreamListener(Sink.INPUT) // destination name 'input.myGroup'
	public void handle(String value) {
		throw new RuntimeException("BOOM!");
	}

	@StreamListener("stava2.myGroup.errors") //channel name 'input.myGroup.errors'
	public void error(Message<?> message) {
		System.out.println("Handling ERROR: " + message);
	}
}

spring.cloud.stream.bindings.input.destination=stava2
spring.cloud.stream.bindings.input.group=myGroup
spring.cloud.stream.bindings.input.content-type=application/json
```

error:
```java
***************************
APPLICATION FAILED TO START
***************************

Description:

A component required a bean named 'stava2.myGroup.errors' that could not be found.


Action:

Consider defining a bean named 'stava2.myGroup.errors' in your configuration.
```

**Comment by olegz** [comment]
@stavalfi It appears there is another issue., For now please use 
```
@ServiceActivator(inputChannel="stava2.myGroup.errors")
```

**Comment by olegz** [comment]
@stavalfi I've identified it as a bug which will be fixed shortly https://github.com/spring-cloud/spring-cloud-stream/issues/1338
Once again, thanks for your persistence! That in itself is an invaluable contribution.

**Comment by garyrussell** [comment]
```java
	@StreamListener("stava2.myGroup.errors") //channel name 'input.myGroup.errors'
	public void error(Message<?> message) {
		System.out.println("Handling ERROR: " + message);
	}
```

That's not supposed to work - `@StreamListener` is for channels bound to external (broker) destinations. The error channel is a local SI channel and can't be used there; use `@ServiceActivator` to consume from that channel.

**Comment by stavalfi** [comment]
@olegz Thanks for all the help. I really appreciate it :) 

For now, don't close this issue yet. I'm still reading your guide.

@garyrussell Thanks for the help. I got a quick question- Why you guys didn't decide that it's not enough to write: `@ServiceActivator(inputChannel = Sink.INPUT +".myGroup".errors")`? I thought that every destination is bounded to a specific channel. What am I missing here?

**Comment by olegz** [comment]
@stavalfi I've already closed it as it was determined that it is the intended behavior and not a bug. We're certainly up for revisiting it in subsequent release, but for now it stands and I've already updated documentation.
As to your question "Why you guys didn't decide that it's not enough to write: @ServiceActivator(inputChannel = Sink.INPUT +".myGroup".errors")". I am not sure what you're asking.  It is enough, right? That is how you annotate application provided error handler.
